### ISSUE TITLE
Replace Mersenne Twister with a modern parallel RNG?

### ISSUE THREAD

**Issue by jgreitemann** [ALPSCore/ALPSCore#203]
I've noticed that the `mpiadapter` and `mc_base` classes initialize the random number generator with seeds that merely differ by the MPI rank. This is a bad practice and can lead to bias in situations where two processes arrive at the same RNG state as they will oversample that particular trajectory in state space. Cf. [Wikipedia](https://en.wikipedia.org/wiki/Mersenne_Twister#Disadvantages), [this thread](http://scicomp.stackexchange.com/questions/23547/parallel-mersenne-twister-for-monte-carlo), and references therein.

While the Mersenne Twister is sort of the gold standard in MC today, it also performs only mediocre in more rigorous statistical tests such as TestU01.

I would propose to switch to a modern RNG such as [PCG](http://www.pcg-random.org/). It offers multiple streams of random numbers which avoids the aforementioned 'bad collisions', and performs better in TestU01 than the Mersenne Twister. I did some preliminary testing and found a speed-up of 40% for the MC estimation of pi and 30% for the 2D Ising example program with `-O3` flags. There also exists an official reference implementation in C++ which ties into the C++11 `random` facilities, so different RNGs can essentially be tested as drop-in replacements. Another RNG that is faster and better than the Mersenne Twister is `xorshift+`. It does not provide multiple streams but features jump functions which allow one to skip a certain number of random numbers within the same sequence in O(1).

Using the new C++ random number facilities would require enabling C++11, thereby raising the required dependencies. I don't know how much priority you place on keeping ALPSCore compatible with old compilers. This is only meant as a suggestion. If you agree, I would be willing to work on switching to the new RNG but would require help in setting up CMake for C++11 as I'm not well-versed in CMake.

**Comment by egull** [comment]
Hi Jonas

C++11 is currently not possible as it is not supported by many of the compilers we care about.

We have delegated our random number generation to boost random. Could a different boost random generator be used for this, or could it be initialized differently to satisfy your requirements?

For your application codes, you can always introduce your own generator (for example PCG or something else) and use this instead of the PRNG we provide. I'm hesitant to replace the default generator at this time unless there is a good reason for it, as this affects everybody's code.

**Comment by galexv** [comment]
[I find it an interesting coincidence that we discussed the same question in a recent group meeting! :)]

I'd think we can introduce a random number generator class as a template parameter to `alps::mcbase` and default it to our `alps::random01` class. Then a user can always provide his/her custom random class.

Just curious: does that 30% performance improvement in Ising simulation come from a less correlated Markov chains or from improvement in RNG speed?

**Comment by aeantipov** [comment]
Template parameter seems a good solution.

As a side note - it might make some sense to have an unsupported repo for extensions of alpscore. This is similar to what Eigen does with http://eigen.tuxfamily.org/dox/unsupported/, which extends the functionality of Eigen by user-provided classes, but is not officially incorporated into the main repo as it is not maintained by the devs of Eigen.

**Comment by egull** [comment]
I'm not aware of a case that shows this correlation in practice (if it did show up we should be very worried), so I suspect it's RNG speed.

**Comment by jgreitemann** [comment]
I agree that in practice this does not bias the simulations. [Brugger et al., Adv. Radio Sci., 12, 75–81, 2014](http://www.adv-radio-sci.net/12/75/2014/ars-12-75-2014.pdf) estimated the collision probability to be astronomically small for the standard Mersenne Twister due to its huge period of 2^19937-1. However, this becomes a real issue with more efficient generator which have periods of e.g. "only" 2^128 (which is still plenty to do MC) on massively parallel systems. The math of this is reminiscent of the infamous [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem) and the collision probability rapidly increases as a function of parallel cores.

The speed-up of the simulation is due to the RNG being more efficient. This is largely a consequence of its state being only a few bytes (hence the shorter period) compared to the bulky 2KB state of the Mersenne Twister, thus reducing memory latency due to improved cache hit rates. I suspect the improvement to be more dramatic on MIC systems where cache is a more scarce commodity. Again, we're not talking orders of magnitude here but it seems to be very low-hanging fruit to me.

As for templating the random generator, I noticed that right now, `alps::random01` bundles the actual RNG with the uniform distribution of boost. Occasionally, one may need other distributions. E.g. in a current DiagMC code I'm working on, I draw momenta from a Gaussian distribution as this will cancel corresponding terms in the detailed balance equations. It is very convenient to be able to use different distributions with the same RNG in C++11 and would be a pain to need to do some Box-Muller scheme or similar to recover that ability. I totally get that you want to keep ALPSCore compatible with older compilers. As Emanuel pointed out, one can always keep an RNG in the application code. I just wanted to mention this use case if you actually intend to redesign this part of ALPSCore.

I hope you don't mind my opening these issues. As I mentioned earlier, I wouldn't mind to get my hands dirty myself. I'm currently in the process of switching my code over to ALPSCore as I think it is a great effort. Many of my colleagues work with makeshift MC 'libraries' which do not scale well. Do you consider ALPSCore ready to be deployed in a production environment?

**Comment by egull** [comment]
I see. In summary: for our generators and with our implementation of initializing generators by rank we are unlikely to ever see a collision. I disagree then with your statement that initializing them by rank is 'bad practice'!

Regarding other distributions: the random01->engine() is public and can be used to initialize another distributions. We should find a better/cleaner way to expose that to the user.

**Comment by jgreitemann** [comment]
Yes, 'bad practice' is maybe a little harsh. But it is easy to forget about this subtlety when switching to a generator with a period that is not as exceptionally large as the one of MT19937. Especially, when templating the RNG, keeping the `seed_offset` mechanism in place as-is would be misleading in that regard.

**Comment by egull** [comment]
OK, so let's put the following on a list of enhancements:

1. introduce a more transparent access to other distributions and the engine.
2. give users a way to change their random number generators to one that they like better, while keeping mt19937 as the default
3. force users to provide a custom initialization along with a custom generator so that they need to think about initializing it in cases where the periods are different.
